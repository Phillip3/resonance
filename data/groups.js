import axios from 'axios';
import * as Amplitude from 'expo-analytics-amplitude';
import config from './config';

const baseURL = 'https://echo-api.westus.cloudapp.azure.com';

// set the featured group, which is sorted to the top of the list of groups
const isFeaturedGroup = (groupName) =>
  groupName.toLowerCase().includes(config.featuredGroup);

export async function getCategories() {
  const { data, status } =
    (await axios.get(`${baseURL}/groups/categories`).catch((err) =>
      Amplitude.logEventWithProperties('ERROR hitting `/groups/categories`', {
        error: err,
      })
    )) || {};

  if (status !== 200) {
    Amplitude.logEventWithProperties('ERROR loading group categories', {
      data,
      status,
    });
    return [];
  }

  const { categories = [] } = data;

  return categories
    .map(({ name }) => {
      if (name === 'Sermon-based') {
        return false;
      }
      if (name === 'Easter') {
        return false;
      }
      if (name === 'Co-Ed') {
        return 'Co-Ed / General';
      }

      return name;
    })
    .filter(Boolean);
}

export async function getOpenGroups() {
  const errorMessage = 'Error fetching groups';
  const { data = [] } = (await axios.get(`${baseURL}/groups/open`)) || {};

  if (!data || !Array.isArray(data)) {
    Amplitude.logEventWithProperties('ERROR loading groups', {
      error: errorMessage,
    });
    throw Error(errorMessage);
  }

  return data.sort(({ name = '' }, { name: otherName = '' }) => {
    if (isFeaturedGroup(name) && !isFeaturedGroup(otherName)) {
      return -1;
    }
    if (!isFeaturedGroup(name) && isFeaturedGroup(otherName)) {
      return 1;
    }
    return 0;
  });
}

export function askQuestion(groupId, firstName, lastName, email, question) {
  const params = {
    GroupID: groupId,
    FirstName: firstName,
    LastName: lastName,
    Email: email,
    Question: question,
  };

  if (__DEV__) {
    console.log({ params });
    return Promise.resolve(true);
  }
  return axios.get(`${baseURL}/groups/ask`, { params });
}

export function joinGroup(groupId, firstName, lastName, email) {
  const params = {
    GroupID: groupId,
    FirstName: firstName,
    LastName: lastName,
    Email: email,
  };

  if (__DEV__) {
    console.log({ params });
    return Promise.resolve(true);
  }
  return axios.get(`${baseURL}/groups/join`, { params });
}
