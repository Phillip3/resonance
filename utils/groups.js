// @flow

export const getCampusCode = (campus: string = '') => {
  switch (campus) {
    case 'NORTH SAN JOSE':
      return 'NSJ';

    case 'SOUTH SAN JOSE':
      return 'SSJ';

    case 'SUNNYVALE':
      return 'SVL';

    case 'FREMONT':
      return 'FMT';

    default:
  }
};

export const getMeetingFrequency = (
  frequency: string = '',
  interval: string = ''
) => {
  if (interval.toLowerCase() === 'weekly') {
    if (frequency === '1') {
      return 'Every week';
    }

    return `Every ${frequency} weeks`;
  }

  if (interval.toLowerCase() === 'monthly') {
    if (frequency === '1') {
      return 'Every month';
    }

    return `Every ${frequency} months`;
  }
};

const getGetOrdinal = (n: number) => {
  const s = ['th', 'st', 'nd', 'rd'];
  const v = n % 100;

  return n + (s[(v - 20) % 10] || s[v] || s[0]);
};

export const getMeetingDay = (
  daysOfWeek: Array<string> = [],
  dayOfMonth: string
) => {
  if (daysOfWeek.length) {
    if (daysOfWeek.length === 7) {
      return 'Every day';
    }

    return daysOfWeek.join(', ');
  }

  if (dayOfMonth) {
    const day = getGetOrdinal(parseInt(dayOfMonth, 10));

    return `the ${day}`;
  }
};

export const getMeetingTime = (time: string = '') => {
  const [hours, minutes] = time
    .split('(')[0]
    .trim()
    .split(':');
  const hrs = parseInt(hours, 10);

  let h = hrs % 12;

  if (h === 0) {
    h = 12;
  }
  return `${h}:${minutes}${hrs < 12 ? 'am' : 'pm'}`;
};
