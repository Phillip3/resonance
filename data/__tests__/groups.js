import axios from 'axios';
import { getCategories } from '../groups';

jest.mock('axios');

test('should get group categories', async () => {
  axios.get.mockResolvedValue({
    status: 200,
    data: {
      categories: [
        {
          name: 'Sermon-based',
          id: '20090',
        },
        {
          name: 'Easter',
          id: '20975',
        },
        {
          name: 'Men Only',
          id: '13299',
        },
        {
          name: 'Women Only',
          id: '13300',
        },
        {
          name: 'Young Professionals',
          id: '13304',
        },
        {
          name: 'Students',
          id: '13305',
        },
        {
          name: 'Recreational / Social',
          id: '15168',
        },
        {
          name: 'Co-Ed',
          id: '15169',
        },
        {
          name: 'Family / Parenting',
          id: '16466',
        },
        {
          name: 'Service / Outreach / Prayer',
          id: '16467',
        },
        {
          name: 'Business / Finance',
          id: '17873',
        },
        {
          name: 'Exploring',
          id: '20152',
        },
        {
          name: 'Marriage',
          id: '21329',
        },
        {
          name: 'Bible Study',
          id: '22396',
        },
      ],
    },
  });

  const categories = await getCategories();

  expect(categories).toMatchSnapshot();
});
