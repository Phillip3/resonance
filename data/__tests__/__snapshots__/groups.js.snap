// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get group categories 1`] = `
Array [
  "Men Only",
  "Women Only",
  "Young Professionals",
  "Students",
  "Recreational / Social",
  "Co-Ed / General",
  "Family / Parenting",
  "Service / Outreach / Prayer",
  "Business / Finance",
  "Exploring",
  "Marriage",
  "Bible Study",
]
`;
